### PR TITLE
Add Pareto frontier filtering for GEPA optimization

### PR DESCRIPTION
**Pull request by anndvision**
Implements and iterates on the Pareto frontier analysis module for GEPA (Genetic Prompt Adaptation), enabling multi-objective prompt optimization across multiple evaluators. The stateful `ParetoFrontier` tracks variants, frequencies, validation scores, cached objective vectors, and the layout fingerprint (datapoint order + metric optimize directions) to support iterative GEPA runs.

## Key Features

- **3-step GEPA filtering**: instance-wise Pareto sets → union → global dominance across (datapoint × evaluator) objectives.
- **Multi-objective support**: mixed maximize/minimize, Pareto dominance (better-or-equal on all, strictly better on ≥1).
- **Missing data handling**: aggressive imputation (-∞ for max, +∞ for min) with high-missing-rate warnings.
- **Validation**: rejects missing datapoints or unknown evaluator keys; enforces deterministic ordering and constant layout for cache correctness.
- **Cache hygiene**: precomputed objective vectors with length checks, pruning of dominated entries; shared optimize-direction extractor to avoid drift.
- **Lightweight improvement check**: `is_improvement` for mutation evaluation on summary stats.

## Implementation Details

- Core logic in `tensorzero-optimizers/src/gepa/pareto.rs` (stateful frontier, dominance helpers, cache).
- Extensive tracing for each pipeline step and missing-data diagnostics.

## Tests

- Unit/integration coverage for dominance scenarios, mixed directions, missing data, cache correctness/pruning, unknown-metric rejection, and performance sanity.
- `cargo test -p tensorzero-optimizers`.

----

> [!IMPORTANT]
> Add Pareto frontier analysis module for GEPA optimization with multi-objective support and extensive testing.
> 
>   - **Behavior**:
>     - Implements `ParetoFrontier` in `pareto.rs` for multi-objective optimization in GEPA.
>     - Supports mixed maximize/minimize objectives and handles missing data with imputation.
>     - Provides `is_improvement` function for mutation evaluation.
>   - **Implementation**:
>     - Core logic in `pareto.rs` includes stateful frontier, dominance helpers, and cache management.
>     - Extensive logging and diagnostics for missing data and cache operations.
>   - **Tests**:
>     - Unit and integration tests for dominance scenarios, mixed directions, missing data, and cache correctness.
>     - Tests added for performance and edge cases in `pareto.rs`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=tensorzero%2Ftensorzero&utm_source=github&utm_medium=referral)<sup> for 9fa6837b9cf9bccea0e4f199c7f30992eeb12c92. You can [customize](https://app.ellipsis.dev/tensorzero/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: tensorzero-optimizers/src/gepa/mod.rs [hunk 1]
```diff
@@ -25,8 +25,10 @@ use tensorzero_core::{
 use crate::{JobHandle, Optimizer};
 
 mod evaluate;
-
+mod pareto;
+// TODO: do not re-export
 pub use evaluate::create_evaluation_dataset;
+pub use pareto::{is_improvement, update_pareto_frontier};
```
Comment (virajmehta) [inline tensorzero-optimizers/src/gepa/mod.rs]: yeah why are we re-exporting?

File: tensorzero-optimizers/src/gepa/pareto.rs [hunk 2]
```diff
@@ -0,0 +1,1760 @@
...
+    error::{Error, ErrorDetails},
+    evaluations::{EvaluatorConfig, InferenceEvaluationConfig},
+    variant::chat_completion::UninitializedChatCompletionConfig,
+};
+
+// Type aliases for cleaner score map signatures (TODO: will live next to evaluation code)
+pub type EvaluatorName = String;
+
+/// Unique identifier for a datapoint/example in a dataset
+pub type DatapointId = String;
```
Comment (virajmehta) [inline tensorzero-optimizers/src/gepa/pareto.rs]: we should be doing more of this in the codebase honestly. One question though is why is DatapointId a string not a Uuid?

**Comment by virajmehta** [comment]
@codex review

File: tensorzero-optimizers/src/gepa/pareto.rs [hunk 3]
```diff
@@ -0,0 +1,2511 @@
...
+    /// Maps variant names to pre-computed flattened objective vectors.
+    /// Vectors are keyed by variant name and validated against the layout fingerprint.
+    /// Private - internal implementation detail for performance optimization.
+    objective_vector_cache: HashMap<String, Arc<Vec<f32>>>,
+
+    /// Datapoint IDs defining the objective space (layout fingerprint)
+    ///
+    /// This ordering must remain constant for cache correctness.
+    /// Private to enforce validation through update methods.
+    datapoint_ids: Vec<String>,
```
Comment (virajmehta) [inline tensorzero-optimizers/src/gepa/pareto.rs]: Uuid?

File: tensorzero-optimizers/src/gepa/pareto.rs [hunk 4]
```diff
@@ -0,0 +1,2511 @@
...
+                        .join(", "),
+                    collisions.len() - 5
+                )
+            };
+            return Err(Error::new(ErrorDetails::InternalError {
+                message: format!(
+                    "Variant name collision(s) detected: {collision_list}. New variants cannot have the same names as existing frontier variants."
+                ),
+            }));
+        }
```
Comment (virajmehta) [inline tensorzero-optimizers/src/gepa/pareto.rs]: I don't mind having fancy error handling like this but please pull it into a separate method with an obvious name so this code is easy to read. Actually, the entire collision detection logic should be pulled out.

File: tensorzero-optimizers/src/gepa/pareto.rs [hunk 5]
```diff
@@ -0,0 +1,2511 @@
...
+pub struct ParetoFrontier {
+    /// Pareto-optimal variants (non-dominated after GEPA's 3-step filtering)
+    pub variants: HashMap<String, UninitializedChatCompletionConfig>,
+
+    /// Instance-wise membership frequencies for weighted sampling
+    ///
+    /// Maps each variant to the count of datapoints where it's Pareto-optimal.
+    /// Higher values indicate the variant performs well across more instances.
+    /// Used for proportional sampling in GEPA's mutation step.
+    pub frequencies: HashMap<String, usize>,
```
Comment (virajmehta) [inline tensorzero-optimizers/src/gepa/pareto.rs]: TODO: make this private and add a sample() associated method

File: tensorzero-optimizers/src/gepa/pareto.rs [hunk 6]
```diff
@@ -0,0 +1,2511 @@
...
+/// This threshold helps identify variants that may be unreliable due to
+/// systematic evaluation failures.
+const HIGH_MISSING_RATE_THRESHOLD: f32 = 0.3;
+
+/// Scores for all evaluators on a single datapoint
+pub type DatapointScores = HashMap<EvaluatorName, Option<f32>>;
+
+/// Scores for all datapoints for a single variant
+pub type VariantScores = HashMap<DatapointId, DatapointScores>;
+
```
[the remark this answers is not in the document]
Comment (virajmehta) [reply tensorzero-optimizers/src/gepa/pareto.rs]: or even `pub struct VariantName(String)` would be more type-safe

File: tensorzero-optimizers/src/gepa/pareto.rs [hunk 7]
```diff
@@ -0,0 +1,2511 @@
...
+                    unexpected.len() - 5
+                )
+            };
+            return Err(Error::new(ErrorDetails::InternalError {
+                message: format!(
+                    "Validation scores contain {} datapoint(s) not in frontier's datapoint_ids: {unexpected_list}. \
+                     Layout must remain constant across iterations.",
+                    unexpected.len()
+                ),
+            }));
```
Comment (virajmehta) [inline tensorzero-optimizers/src/gepa/pareto.rs]: pull this out

File: tensorzero-optimizers/src/gepa/pareto.rs [hunk 8]
```diff
@@ -0,0 +1,2511 @@
...
+
+            if !has_any_score {
+                tracing::warn!(
+                    "All evaluation scores are None - no variant produced any valid scores"
+                );
+                return Err(Error::new(ErrorDetails::InternalError {
+                    message: "All variants failed to produce valid scores".to_string(),
+                }));
+            }
+        }
```
Comment (virajmehta) [inline tensorzero-optimizers/src/gepa/pareto.rs]: pull this out

File: tensorzero-optimizers/src/gepa/pareto.rs [hunk 9]
```diff
@@ -0,0 +1,2511 @@
...
+        );
+
+        // Step 1: Build instance-wise Pareto sets P*[i] for each datapoint
+        // Original paper: P*[i] = {variants with max score on instance i} (single evaluator)
+        // Our extension: P*[i] = {Pareto non-dominated variants on instance i} (multiple evaluators)
+        let mut instance_pareto_sets: HashMap<String, HashSet<String>> = HashMap::new();
+
+        for datapoint_id in &self.datapoint_ids {
+            // Collect scores for this datapoint across all variants
+            let mut instance_scores: Vec<(String, HashMap<String, Option<f32>>)> = Vec::new();
```
Comment (virajmehta) [inline tensorzero-optimizers/src/gepa/pareto.rs]: use some of the new types to make it obvious what the data structure here is; it is not clear from inspection.

File: tensorzero-optimizers/src/gepa/pareto.rs [hunk 10]
```diff
@@ -0,0 +1,2329 @@
...
+                        .join(", "),
+                    collisions.len() - 5
+                )
+            };
+            return Err(Error::new(ErrorDetails::InternalError {
+                message: format!(
+                    "Variant name collision(s) detected: {collision_list}. New variants cannot have the same names as existing frontier variants."
+                ),
+            }));
+        }
```
Comment (virajmehta) [inline tensorzero-optimizers/src/gepa/pareto.rs]: pull this out

File: tensorzero-optimizers/src/gepa/pareto.rs [hunk 11]
```diff
@@ -0,0 +1,2329 @@
...
+                    let keep = self.optimize_directions.contains_key(metric);
+                    if !keep {
+                        dropped_metrics.insert(metric.clone());
+                    }
+                    keep
+                });
+            }
+
+            new_variant_scores.insert(name, scores);
+        }
```
Comment (virajmehta) [inline tensorzero-optimizers/src/gepa/pareto.rs]: pull this out

File: tensorzero-optimizers/src/gepa/pareto.rs [hunk 12]
```diff
@@ -0,0 +1,2329 @@
...
+                    .join(", ")
+            );
+        }
+
+        tracing::debug!(
+            "Combined candidates: {} total ({} existing + {} new)",
+            all_candidates.len(),
+            self.variants.len(),
+            new_variant_scores.len()
+        );
```
Comment (virajmehta) [inline tensorzero-optimizers/src/gepa/pareto.rs]: pull this out

**Review by virajmehta — CHANGES_REQUESTED** [review]
`update` needs to be factored out a lot. Please break it up and document each piece.

File: tensorzero-optimizers/src/gepa/mod.rs (hunk 1 above)
Comment (anndvision) [reply tensorzero-optimizers/src/gepa/mod.rs]: temporary to avoid unused flags

**Review by virajmehta — APPROVED** [review]
